docs/rfcs/generic-record-sync.md

```diff
@@ -0,0 +1,1287 @@
+fix up the set of stale guids in any collection that contains `ForeignGuid`'s
+after syncing.
+
+Locally, multiple collections are stored in a single database, and every
+collection has both the local copy and mirror available, so that a 3-way-merge
+may be performed.
+
+## The Schema
+
+Note: This has a bunch of things in it we don't need to support in the initial
+version. In particular, the initial version should absolutely drop
+`ForeignGuid`, `RecordSet`, and `UntypedMap`. All of these, IMO, are important
+to the overall usability story here, however, our plan allows for adding new
+field types/merge strategies here without locking out clients that don't use
+them.
+
+Additionally, it's worth noting that the decision not to add rich types here is
+intentional. This is a representation of how the data is stored and merged, and
+that's more or less all. (Note: now that we have a plan for adding new types to
+this set I feel a lot less strongly here, but none-the-less)
+
+I've written up rust code to describe the schema, which is at the bottom of the
+document because it's fairly long due to many comments.
+
+In practice, I belive schemas would look something like:
+
+```json
+{
+    // sync 1.5 collection name
+    "name": "passwords",
+    "version": "0.1.0",
+
+    // Optional, and defaults to the minimum version
+    // still semver-compatible with 'version'
+    "required_version": "0.1.0",
+    "legacy": true,
+    "dedupe_on": ["hostname", "username", "formSubmitURL", "httpRealm"],
+    "fields": [
+        { "name": "id", "type": "own_guid" },
+        { "name": "hostname", "type": "text" },
+
+        { "name": "formSubmitURL", "type": "text" },
+        { "name": "httpRealm", "type": "text" },
+
+        { "name": "username", "type": "text" },
+        { "name": "password", "type": "text" },
+
+        // In practice I don't know if these are actually deprecated.
+        { "name": "usernameField", "type": "text", "deprecated": true },
+        { "name": "passwordField", "type": "text", "deprecated": true },
+
+        { "name": "timeCreated", "type": "number", "merge": "take_min", "default": 0 },
+        { "name": "timePasswordChanged", "type": "number", "merge": "take_max", "default": 0 },
+        { "name": "timeLastUsed", "type": "number", "merge": "take_max", "default": 0 },
+        { "name": "timesUsed",  "type": "number", "merge": "take_sum", "default": 0 }
+    ]
+}
+```
+
+See the comments in the rust code rust code for a description of what this would
+mean, what restrictions we'd impose, etc.
+
+## Guid Renames
+
+In order to support references to other IDs, we primarially need to handle
+deduping. E.g, the guid on the server may be what a record used to be called,
+however now it has a differnt name. My recommendation is a couple small changes
+to the server protocol to support it, however we could also implement this
+ourselves with more difficulty (and lack of transactional semantics).
+
+### Server support
```

Comment:
Changed my mind here, clients should just sync these. we loose atomicity but thats not required for this.

Reply:
As noted elsewhere, foreign_guid is dead until an actual use case exists. It was arguably the most complex thing in this (requiring changes to the server, and all existing clients), was guaranteed not to work perfectly, and could easily have never been something anybody wanted to use.